File: plugin/src/main/java/de/infolektuell/gradle/jextract/model/JextractDataStore.java
package de.infolektuell.gradle.jextract.model;

import de.infolektuell.gradle.jextract.model.DownloadClient.Resource;

import java.io.InputStream;
import java.net.URI;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Properties;

/// Helpers for consistent paths and filenames for Jextract, depending on version and platform
public class JextractDataStore {
    /// Tries to create an instance of JextractDataStore using data from a properties file
    /// @param path The file that contains the Jextract distribution data to be loaded
    /// @return a JextractDataStore instance
    public static JextractDataStore create(Path path) {
        try (var f = Files.newBufferedReader(path)) {
            var data = new Properties();
            data.load(f);
            return new JextractDataStore(data);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    /// Tries to create an instance of JextractDataStore using the default distribution data
    /// @return a JextractDataStore instance
    public static JextractDataStore create() {
        try (InputStream s = JextractDataStore.class.getResourceAsStream("/jextract.properties")) {
            var data = new Properties();
            data.load(s);
            return new JextractDataStore(data);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private final Platform platform = Platform.getCurrentPlatform();
    private final Properties data;
    JextractDataStore(Properties data) {
        this.data = data;
    }

    /// Constructs the file name of the Jextract executable depending on the current operating system
    /// @return The constructed file name
    public String getExecutableFilename() {
        return platform.isWindows() ? "jextract.bat" : "jextract";
    }

    /// Finds the matching Jextract version for a given Java major version
    /// @param javaVersion A major Java language version
    /// @return The most appropriate Jextract version
    public int version(int javaVersion) {
        if (javaVersion == 25) {
            return 25;
        }
        return Integer.max(Integer.min(javaVersion, 22), 19);
    }

    /// Creates a downloadable resource for the specified Jextract version
    /// @param javaVersion A major Java language version
    /// @return The constructed resource matching the java version
    public Resource resource(int javaVersion) {
        int version = version(javaVersion);
        String os = platform.operatingSystem().name().toLowerCase();
        String arch = platform.architecture().name().toLowerCase();
        String base = String.format("jextract.%d.%s.%s", version, os, arch);
        String baseFallback = String.format("jextract.%d.%s.%s", version, os, "x64");
        String url = data.getProperty(base + ".url", data.getProperty(baseFallback + ".url"));
        String checksum = data.getProperty(base + ".sha-256", data.getProperty(baseFallback + ".sha-256"));
        return new Resource(URI.create(url), checksum, "SHA-256");
    }

    /// Constructs an archive filename for the specified Jextract version
    /// @param javaVersion A major Java language version
    /// @return the constructed file name
    public String filename(int javaVersion) {
        int version = version(javaVersion);
        String os = platform.operatingSystem().name().toLowerCase();
        String arch = platform.architecture().name().toLowerCase();
        String base = String.format("jextract.%d.%s.%s", version, os, arch);
        String baseFallback = String.format("jextract.%d.%s.%s", version, os, "x64");
        String url = data.getProperty(base + ".url", data.getProperty(baseFallback + ".url"));
        var segments = url.split("/");
        return segments[segments.length - 1];
    }
}

File: plugin/src/main/kotlin/de/infolektuell/gradle/jextract/model/JextractDataStore.kt
package de.infolektuell.gradle.jextract.model

import de.infolektuell.gradle.jextract.model.DownloadClient.Resource
import java.net.URI
import java.nio.file.Files
import java.nio.file.Path
import java.util.*
import kotlin.math.max
import kotlin.math.min

/** A service class to query data for downloading Jextract, depending on the current platform and a given target JVM version */
class JextractDataStore {

    private val platform = Platform.getCurrentPlatform()
    private val os get() = platform.operatingSystem.name.lowercase()
    private val arch get() = platform.architecture.name.lowercase()
    private val distributionCache = mutableMapOf<Path, Properties>()
    private val defaultDistributions: Properties by lazy { loadDefaultDistributions() }
    val executableFilename: String by lazy { if (platform.isWindows) "jextract.bat" else "jextract" }

    /** Returns the matching Jextract version for a given [Java major version][javaVersion] */
    @Suppress("SameParameterValue")
    fun version(javaVersion: Int, lower: Int = 19, upper: Int = 22) = max(min(javaVersion, upper), lower)

    /** Returns a downloadable resource for the specified Jextract [version],  using the data from an optional [file], falls back to the official distribution data */
    fun resource(version: Int, file: Path? = null) = resource(version, loadDistributions(file))

    /** Returns an archive filename for the specified Jextract [version],  using the data from an optional [file], falls back to the official distribution data */
    fun filename(version: Int, file: Path? = null) = filename(version, loadDistributions(file))

    private fun resource(javaVersion: Int, data: Properties): Resource {
        val version = version(javaVersion)
        val url: String = data.getProperty("jextract.$version.$os.$arch.url") ?: data.getProperty("jextract.$version.$os.x64.url")
        val checksum: String = data.getProperty("jextract.$version.$os.$arch.sha-256") ?: data.getProperty("jextract.$version.$os.x64.sha-256")
        return Resource(URI.create(url), checksum)
    }

    private fun filename(javaVersion: Int, data: Properties): String {
        val version = version(javaVersion)
        val url: String = data.getProperty("jextract.$version.$os.$arch.url") ?: data.getProperty("jextract.$version.$os.x64.url")
        return url.replaceBeforeLast('/', "").trim('/')
    }

    private fun loadDistributions(file: Path? = null): Properties {
        if (file == null) return defaultDistributions
        return distributionCache.computeIfAbsent(file) { k ->
            Properties(defaultDistributions).apply {
                load(Files.newBufferedReader(k))
            }
        }
    }

    private fun loadDefaultDistributions(): Properties {
        return Properties().apply {
            object {}.javaClass.getResourceAsStream("/jextract.properties")?.use{
                load(it)
            }
        }
    }
}
